rancher/rancher
# [WIP] Redirect ServiceMonitor scraping to cluster-level Prometheus

**Problem:**
Cluster-level Prometheus could not scrape the ServiceMonitors lied on non-System project.
This is not friendly to `prometheus-adapter` and `thanos`.

**Solution:**
Make cluster-level Prometheus could scrape the ServiceMonitors from
all project-level.

**Related PR:**
https://github.com/rancher/system-charts/pull/119

**After using this PR:**
1. It would not affect the previous running state of the project-level monitoring and cluster-level.
1. If the users have a requirement for using the project-level monitoring in the future, they need to upgrade the `cluster-monitoring` app.
1. If the users only enable project-level monitoring (only consider the Custom Metrics) without cluster-level, they must enable the cluster-level monitoring and then could upgrade the `project-monitoring` app. (After upgrading the `project-monitoring` app, the project-level monitoring could not take over its project's ServiceMonitors, but the cluster-level would)
1. **warning**: With `project-monitoring` and `cluster-monitoring` apps upgrading, the project-level might get duplicate Custom Metric data in a period. The period depends on the configuration refreshing speed of the Prometheus Operator, for example:
    1. Before:
         ```
         # project-level Prometheus scraped
         web_app_online_user_count{}
         ```
    1. Transition (~ 5 mins):
         ```
         # project-level Prometheus scraped
         web_app_online_user_count{}
         # project-level Prometheus federated from cluster-level
         web_app_online_user_count{prometheus="cattle-prometheus/cluster-monitoring",prometheus_from="test",prometheus_replica="prometheus-cluster-monitoring-0}
         ```
    1. After: 
        ```
         # project-level Prometheus federated from cluster-level
         web_app_online_user_count{prometheus="cattle-prometheus/cluster-monitoring",prometheus_from="test",prometheus_replica="prometheus-cluster-monitoring-0}
        ```

1. ServiceMonitor & PrometheusRule scraping target:
    1. Before:

        | Level | ServiceMonitor | PrometheusRules |
        | :---: | :---: | :---: |
        | cluster | `System` project | `System` project |
        | project | owned project | owned project |

    1. After:

        | Level | ServiceMonitor | PrometheusRules |
        | :---: | :---: | :---: |
        | cluster | all project | `System` project |
        | project | --- | owned project |

**Test plan:**
1. Setup a cluster with the previous version Rancher server that allowed project-level monitoring
1. Enable cluster-level and one project-level monitoring
1. Configure some Custom Metrics under that project
1. Upgrade the version included this PR, wait for a while:
    - [ ] [Verify] the project-level Prometheus should never restart
    - [ ] [Verify] the cluster-level Prometheus should never restart
    - [ ] [Verify] the Alert Rules could only query from project-level Prometheus
    - [ ] [Verify] the Custom Metrics could only query from project-level Grafana
1. Upgrade the `cluster-monitoring` and `project-monitoring` apps, wait for a while:
    - [ ] [Verify] the project-level Prometheus should never restart
    - [ ] [Verify] the cluster-level Prometheus should never restart
    - [ ] [Verify] the Alert Rules could query from project-level Prometheus
    - [ ] [Verify] the Custom Metrics could query from both project-level and cluster-level Grafana
